[MRG] estimated step size for SAG* objects

fabianp opened the pull request:
This estimates the step size from the data instead of relying on the user having to specify a step size. The result is that it converges for many cases for which the default step size will not converge, such as [multiplying the 20news dataset by 3](https://gist.github.com/fabianp/1c80fc522617f3d08f6d).

![](http://fa.bianp.net/tmp/2015/bench_img.png)

Some remarks:

* The modified Huber loss is not Lipschitz-smooth so the step size is not well defined (AFAIK the theory does not guarantee convergence in this case). The step size that we propose in this case is given taking a bound on the second derivative ignoring the discontinuities, which seems to work fine in practice.

* The hard work is done by @zermelozf, I just did cosmetic changes.

mblondel commented on lightning/impl/sag.py:
```diff
@@ -1,8 +1,20 @@
+    def get_max_squared_sum(X):
+        n = X.shape[0]
+        return np.max(X.dot(X.T)[range(n), range(n)])
+
```
I would use `np.sum(X ** 2, axis=1).max()` in the dense case and `csr_row_norms(X).max()` in the CSR case (from sklearn.utils.sparsefuncs).

fabianp replied:
I think that function is available in [scikit-learn 0.14](https://github.com/scikit-learn/scikit-learn/tree/0.14.X/sklearn/utils), it looks like it was added in 0.15

fabianp replied:
I used your formula for the dense case at least.

mblondel replied:
We can bump the dependency to 0.15 (released in July 2014).

mblondel commented:
Very nice! Thanks!

fabianp commented:
I've changed the backport of get_max_squared_sum to use csr_row_norms as suggested by @mblondel . I've also updated the README to reflect this (the import is anyway lazy so lightning should at least import in older versions of sklearn).

mblondel commented:
Merging, thank you very much!
We can add this for SVRG too but this can be done in another PR.
